Framework/DataHandling/src/SaveGSS.cpp

```diff
@@ -92,391 +136,502 @@ void SaveGSS::init() {
+    const std::string error = strerror(errno);
+    g_log.error("Failed to open file. Error was: " + error);
+    throw std::runtime_error("Could not open the file at the following path: " +
+                             outFilePath);
+  }
+  // Stream is good - return to caller through a std::move
+  return std::move(outStream);
+}
+
+/** Ensures that when a workspace group is passed as output to this workspace
+*  everything is saved to one file and the bank number increments for each
+*  group member.
+*  @param alg ::           Pointer to the algorithm
+*  @param propertyName ::  Name of the property
+*  @param propertyValue :: Value  of the property
+*  @param periodNum ::     Effectively a counter through the group members
+*/
+void SaveGSS::setOtherProperties(IAlgorithm *alg,
+                                 const std::string &propertyName,
+                                 const std::string &propertyValue,
+                                 int periodNum) {
+  // We want to append subsequent group members to the first one
+  if (propertyName == "Append") {
+    if (periodNum != 1) {
+      alg->setPropertyValue(propertyName, "1");
+    } else
+      alg->setPropertyValue(propertyName, propertyValue);
+  }
+  // We want the bank number to increment for each member of the group
+  else if (propertyName == "Bank") {
+    alg->setProperty("Bank", std::stoi(propertyValue) + periodNum - 1);
+  } else
+    Algorithm::setOtherProperties(alg, propertyName, propertyValue, periodNum);
 }
 
 /**
+  * Validates the user input is matches certain constraints
+  * in length and type and logs a warning or throws depending on
+  * whether we can continue.
+  *
+  * @throws:: If for any reason we cannot run the algorithm
   */
-void SaveGSS::writeRALFdata(const int bank, const bool MultiplyByBinWidth,
-                            std::stringstream &out,
-                            const HistogramData::Histogram &histo) const {
+void SaveGSS::validateUserInput() const {
+  // Check whether it is PointData or Histogram
+  if (!m_inputWS->isHistogramData())
+    g_log.warning("Input workspace is NOT histogram! SaveGSS may not work "
+                  "well with PointData.");
 
-  const auto &xVals = histo.x();
-  const auto &yVals = histo.y();
-  const auto &eVals = histo.e();
-  const size_t datasize = yVals.size();
+  // Check the number of histogram/spectra < 99
+  const int nHist = static_cast<int>(m_inputWS->getNumberHistograms());
+  if (nHist > 99) {
+    std::string outError = "Number of Spectra(" + std::to_string(nHist) +
+                           ") cannot be larger than 99 for GSAS file";
+    g_log.error(outError);
+    throw std::invalid_argument(outError);
+  }
+
+  // Check we have any output filenames
+  assert(m_outFileNames.size() > 0);
+
+  const std::string filename = getProperty("Filename");
+  const bool append = getProperty("Append");
+  for (const auto &filename : m_outFileNames) {
```

Comment:
clang-tidy warning: `Local variable 'filename' hides previous declaration in the upper scope`